test/type-parser.js

```diff
@@ -0,0 +1,237 @@
+          nodeType: 'h1',
+          message: 'duplicate field name "[[A]]"',
+        }
+      );
+    });
+
+    it('ambiguous or', async () => {
+      await assertError(
+        positioned`
+        <emu-clause id="example" type="abstract operation">
+        <h1>
+          ExampleAlg (): a List of null ${M}or Numbers
+        </h1>
+        <dl class="header">
+        </dl>
+        <emu-alg>
+          1. Return NormalCompletion(0).
+        </emu-alg>
+        </emu-clause>
+      `,
+        {
+          ruleId: 'type-parsing',
+          nodeType: 'h1',
+          message:
+            'type is ambiguous; can\'t tell where the "or" attaches (add "either" to disambiguate)',
+        }
+      );
+
+      await assertError(
+        positioned`
+        <emu-clause id="example" type="abstract operation">
+        <h1>
+          ExampleAlg (): ~empty~ or a List of null ${M}or Numbers
+        </h1>
+        <dl class="header">
+        </dl>
+        <emu-alg>
+          1. Return NormalCompletion(0).
+        </emu-alg>
+        </emu-clause>
+      `,
+        {
+          ruleId: 'type-parsing',
+          nodeType: 'h1',
+          message: 'could not determine how to associate "or"; try adding "either"',
+        }
+      );
+    });
+
+    it('error in parameter', async () => {
+      await assertError(
+        positioned`
+          <emu-clause id="example" type="abstract operation">
+          <h1>
+            ExampleAlg (
+              x: a normal completion containing${M},
+              y: unknown,
+            )
+          </h1>
+          <dl class="header">
+          </dl>
+          <emu-alg>
+            1. Return NormalCompletion(0).
+          </emu-alg>
+          </emu-clause>
+        `,
+        {
+          ruleId: 'type-parsing',
+          nodeType: 'h1',
+          message: 'expected to find at type, got empty string',
+        }
+      );
+    });
+  });
+
+  describe('parse trees', () => {
+    it('simple cases', () => {
+      assert.deepStrictEqual(TypeParser.parse(`~unused~`), {
+        kind: 'unused',
+      });
+
+      assert.deepStrictEqual(TypeParser.parse(`~empty~`), {
+        kind: 'opaque',
+        type: '~empty~',
+      });
+
+      assert.deepStrictEqual(TypeParser.parse(`X or Y`), {
+        kind: 'union',
+        types: [
+          {
+            kind: 'opaque',
+            type: 'X',
+          },
+          {
+            kind: 'opaque',
+            type: 'Y',
+          },
+        ],
+      });
+
+      assert.deepStrictEqual(TypeParser.parse(`either X or Y`), {
+        kind: 'union',
+        types: [
+          {
+            kind: 'opaque',
+            type: 'X',
+          },
+          {
+            kind: 'opaque',
+            type: 'Y',
+          },
+        ],
+      });
+
+      assert.deepStrictEqual(TypeParser.parse(`a List`), {
+        kind: 'list',
+        elements: null,
+      });
+
+      assert.deepStrictEqual(TypeParser.parse(`a List of X`), {
```

Comment:
I'd like to see "a List of Lists of X" or similar. Also, a test that covers "records with fields".